File: src/Algorithms/dfs.js
export function dfs(grid, startNode, endNode) {
    const visitedNodesInOrder = [];
    startNode.distance = 0;
    const stack = [startNode];
    while (stack.length) {
        const currentNode = stack.pop();
        if (currentNode.isWall) continue;
        if (currentNode.distance === Infinity) return visitedNodesInOrder;
        currentNode.isVisited = true;
        visitedNodesInOrder.push(currentNode);
        if (currentNode === endNode) return visitedNodesInOrder;
        const unvisitedNeighbors = getUnvisitedNeighbors(currentNode, grid);
        for (const neighbor of unvisitedNeighbors) {
            neighbor.distance = currentNode.distance + 1;
            neighbor.previousNode = currentNode;
            stack.push(neighbor);
        }
    }
}

function getUnvisitedNeighbors(node, grid) {
    const neighbors = [];
    const { col, row } = node;
    if (row > 0) neighbors.push(grid[row - 1][col]);
    if (row < grid.length - 1) neighbors.push(grid[row + 1][col]);
    if (col > 0) neighbors.push(grid[row][col - 1]);
    if (col < grid[0].length - 1) neighbors.push(grid[row][col + 1]);
    return neighbors.filter(neighbor => !neighbor.isVisited);
}

File: golang-src/algorithms/dfs.go
package algorithms

import (
	"math"
	"server/maze"
)

// DFSAlgorithm performs a depth-first search on the grid
func DFSAlgorithm(grid [][]maze.Node, startNode, endNode *maze.Node) []maze.Node {
	visitedNodesInOrder := []maze.Node{}
	startNode.Distance = 0
	stack := []*maze.Node{startNode}

	for len(stack) > 0 {
		currentNode := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if currentNode.IsWall {
			continue
		}
		if currentNode.Distance == math.MaxInt32 { // Equivalent to Infinity
			return visitedNodesInOrder
		}
		currentNode.IsVisited = true
		visitedNodesInOrder = append(visitedNodesInOrder, *currentNode)

		if currentNode == endNode {
			return visitedNodesInOrder
		}

		unvisitedNeighbors := getUnvisitedNeighbors(currentNode, grid)
		for _, neighbor := range unvisitedNeighbors {
			neighbor.Distance = currentNode.Distance + 1
			neighbor.PreviousNode = currentNode
			stack = append(stack, neighbor)
		}
	}

	return visitedNodesInOrder
}
